wpilibj/src/main/java/edu/wpi/first/wpilibj/experimental/controller/PIDController.java

```diff
@@ -0,0 +1,612 @@
+    }
+  }
+
+  /**
+   * Returns the change in error per second.
+   */
+  public double getDeltaError() {
+    double error = getError();
+
+    m_thisMutex.lock();
+    try {
+      return (error - m_prevError) / getPeriod();
+    } finally {
+      m_thisMutex.unlock();
+    }
+  }
+
+  /**
+   * Calculates the output of the PID controller.
+   *
+   * @param measurement The current measurement of the process variable.
+   * @return The controller output.
+   */
+  @SuppressWarnings("LocalVariableName")
+  public double calculate(double measurement) {
+    // Storage for function inputs
+    final double Kp;
+    final double Ki;
+    final double Kd;
+    final double minimumOutput;
+    final double maximumOutput;
+
+    // Storage for function input-outputs
+    double totalError;
+
+    m_thisMutex.lock();
+    try {
+      Kp = m_Kp;
+      Ki = m_Ki;
+      Kd = m_Kd;
+      minimumOutput = m_minimumOutput;
+      maximumOutput = m_maximumOutput;
+
+      m_prevError = m_currError;
+      m_currError = getContinuousError(m_reference - measurement);
+      totalError = m_totalError;
+    } finally {
+      m_thisMutex.unlock();
+    }
+
+    if (Ki != 0) {
+      totalError = clamp(totalError + m_currError * getPeriod(), minimumOutput / Ki,
+          maximumOutput / Ki);
+    }
+
+    double output =
+        clamp(Kp * m_currError + Ki * totalError + Kd * (m_currError - m_prevError) / getPeriod(),
+            minimumOutput, maximumOutput);
+
+    m_thisMutex.lock();
+    try {
+      m_totalError = totalError;
+      m_output = output;
+    } finally {
+      m_thisMutex.unlock();
+    }
+
+    return output;
+  }
+
+  /**
+   * Read the input, calculate the output accordingly, and return the output.
+   *
+   * @param measurement The current measurement of the process variable.
+   * @param reference   The reference (setpoint) of the controller.
+   * @return The controller output.
+   */
+  public double calculate(double measurement, double reference) {
+    setReference(reference);
+    return calculate(measurement);
+  }
+
+  /**
+   * Reset the previous error, the integral term, and disable the controller.
+   */
+  public void reset() {
+    m_thisMutex.lock();
+    try {
+      m_prevError = 0;
+      m_totalError = 0;
+      m_output = 0;
+    } finally {
+      m_thisMutex.unlock();
+    }
+  }
+
+  @Override
+  public final synchronized String getName() {
+    return m_name;
+  }
+
+  @Override
+  public final synchronized void setName(String name) {
+    m_name = name;
+  }
+
+  @Override
+  public final synchronized String getSubsystem() {
+    return m_subsystem;
+  }
+
+  @Override
+  public final synchronized void setSubsystem(String subsystem) {
+    m_subsystem = subsystem;
+  }
+
+  @Override
+  public void initSendable(SendableBuilder builder) {
+    builder.setSmartDashboardType("PIDController");
+    builder.setSafeState(this::reset);
+    builder.addDoubleProperty("Kp", this::getP, this::setP);
```

Comment:
These properties should be renamed to "p", "i", "d", "setpoint" to be compatible with the existing PID controller sendable format. Though with no feedforward constant, it won't be compatible with shuffleboard (as-is, anyway). I might be able to work around that on my end.